### PR TITLE
fix: ListObjects returning IDs not from the chosen type when contextu…

### PR DESCRIPTION
**Pull request by miparnisari**
Fix for https://github.com/openfga/openfga/issues/354

### CODE REVIEW

File: server/commands/list_objects.go [hunk 1]
```diff
@@ -238,26 +238,27 @@ func (q *ListObjectsQuery) performChecks(
...
-	iter1 := storage.NewTupleKeyObjectIterator(req.GetContextualTuples().GetTupleKeys())
+	iter1 := storage.NewTupleKeyObjectIterator(
+		req.GetContextualTuples().GetTupleKeys(),
+		func(tk *openfgapb.TupleKey) bool {
+			return tuple.GetType(tk.GetObject()) == req.GetType()
+		})
 
 	iter2, err := q.Datastore.ListObjectsByType(ctx, req.GetStoreId(), req.GetType())
 	if err != nil {
+		iter1.Stop()
```
Comment (craigpastro) [inline server/commands/list_objects.go:249]: Should this be `defer`ed after its definition?

**Review by craigpastro — APPROVED** [review]
Left one comment, but LGTM.

File: server/commands/list_objects.go (hunk 1 above)
Comment (miparnisari) [reply server/commands/list_objects.go:249]: I don't think so - under normal conditions, the iterator will be stopped in line 256. This is just for the error case, where I need to clean up manually

Comment (craigpastro) [reply server/commands/list_objects.go:249]: Ah, I see. Thanks!

File: storage/storage.go [hunk 2]
```diff
@@ -185,13 +185,16 @@ func NewTupleKeyIteratorFromTupleIterator(iter TupleIterator) TupleKeyIterator {
 }
 
 // NewTupleKeyObjectIterator returns an ObjectIterator that iterates over the objects
-// contained in the provided list of TupleKeys.
-func NewTupleKeyObjectIterator(tupleKeys []*openfgapb.TupleKey) ObjectIterator {
+// contained in the provided list of TupleKeys. You can optionally pass in a function that
+// will remove the objects that don't match the filter.
+func NewTupleKeyObjectIterator(tupleKeys []*openfgapb.TupleKey, filterFunc TupleKeyFilterFunc) ObjectIterator {
```
Comment (jon-whit) [inline storage/storage.go:190]: Please don't modify this signature. We have the [`FilteredTupleKeyIterator`](https://github.com/openfga/openfga/blob/main/storage/storage.go#L261) that serves this same purpose.

There's an example of how to do it here https://github.com/openfga/openfga/blob/main/storage/storage_test.go#L194

Comment (miparnisari) [reply storage/storage.go:190]: The problem is that `FilteredTupleKeyIterator` returns a `TupleKeyIterator`, but in line 256 I need a `ObjectIterator`. 

I could add:

```go
func NewObjectIteratorFromTupleKeyIterator(iter TupleKeyIterator) ObjectIterator {
	objects := make([]*openfgapb.Object, 0)
	for {
		tk, err := iter.Next()
		if err == ErrIteratorDone {
			break
		}
		objectType, objectID := tuple.SplitObject(tk.GetObject())
		objects = append(objects, &openfgapb.Object{Type: objectType, Id: objectID})
	}

	return NewStaticObjectIterator(objects)
}
```

and then in ListObjects
```go
iter1 := storage.NewObjectIteratorFromTupleKeyIterator(storage.NewFilteredTupleKeyIterator(
		storage.NewStaticTupleKeyIterator(req.GetContextualTuples().GetTupleKeys()),
		func(tk *openfgapb.TupleKey) bool {
			return tuple.GetType(tk.GetObject()) == req.GetType()
		}))
iter2 := q.Datastore.ListObjectsByType(ctx, req.GetStoreId(), req.GetType())

iter := storage.NewUniqueObjectIterator(iter1, iter2)
```

Do you prefer it this way? I don't have a strong opinion.

Comment (jon-whit) [reply storage/storage.go:190]: Yeah, I like that chaining approach better as it is more consistent with what we've been doing with the other iterator chaining.

Comment (miparnisari) [reply storage/storage.go]: Updated!

File: storage/storage.go [hunk 3]
```diff
@@ -197,6 +198,24 @@ func NewTupleKeyObjectIterator(tupleKeys []*openfgapb.TupleKey) ObjectIterator {
 	return NewStaticObjectIterator(objects)
 }
 
+func NewObjectIteratorFromTupleKeyIterator(iter TupleKeyIterator) ObjectIterator {
```
Comment (jon-whit) [inline storage/storage.go:201]: Can we please implement this in a way that doesn't require the extra memory to statically store the list of objects?

Instead, we should implement the `ObjectIterator` interface with a struct that uses the `TupleKeyIterator` and yields the objects from that iterator dynamically. For example,

```
type tupleKeyObjectIterator struct {
    iter TupleKeyIterator
}

var _ ObjectIterator = &tupleKeyObjectIterator{} // make sure we implement the ObjectIterator interface

func (t *tupleKeyObjectIterator) Next() (*openfgapb.Object, error) {
    tk, err := t.iter.Next()
    if err != nil {
        return err
    }

    return tk.GetObject(), nil
}
```
then this function simply becomes
```
func NewObjectIteratorFromTupleKeyIterator(iter TupleKeyIterator) ObjectIterator {
    return &tupleKeyObjectIterator{iter}
}
```

**Review by jon-whit — CHANGES_REQUESTED** [review]
https://github.com/openfga/openfga/pull/366/files#r1037342734

File: storage/storage.go (hunk 3 above)
Comment (miparnisari) [reply storage/storage.go]: Good idea. Thanks!

**Review by jon-whit — APPROVED** [review]

**Pull request merged**